build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.cache-miss-monitor.gradle.kts

```diff
@@ -14,31 +14,53 @@
  * limitations under the License.
  */
 
+@file:Suppress("UNCHECKED_CAST")
+
 import com.gradle.scan.plugin.BuildScanExtension
-import java.util.concurrent.atomic.AtomicBoolean
+import org.gradle.kotlin.dsl.support.serviceOf
+
+val gradleRootProject = when {
+    project.name == "gradle" -> project.rootProject
+    project.rootProject.name == "build-logic" -> rootProject.gradle.parent!!.rootProject
+    else -> project.gradle.parent!!.rootProject
+}
 
-val cacheMissTagged = AtomicBoolean(false)
-plugins.withId("com.gradle.build-scan") {
-    val buildScan = extensions.findByType<BuildScanExtension>()
+val rootProjectName = rootProject.name
+val isInBuildLogic = rootProjectName == "build-logic"
 
-    if (System.getenv("TEAMCITY_VERSION") != null) {
-        gradle.taskGraph.afterTask {
-            if (buildCacheEnabled() && isCacheMiss() && isNotTaggedYet()) {
-                buildScan?.tag("CACHE_MISS")
+if (buildCacheEnabled() && System.getenv("TEAMCITY_VERSION") != null) {
+    gradle.taskGraph.whenReady {
+        val cacheMissMonitorBuildService: Provider<CacheMissMonitorBuildService> = gradle.sharedServices.registerIfAbsent("cacheMissMonitorBuildService-$rootProjectName", CacheMissMonitorBuildService::class) {
+            parameters.buildCacheEnabled.set(buildCacheEnabled())
+            parameters.monitoredTaskPaths.set(allTasks.filter { it.isCacheMissMonitoredTask() }.map { if (isInBuildLogic) ":build-logic${it.path}" else it.path }.toSet())
+        }
+        gradle.serviceOf<BuildEventsListenerRegistry>().onTaskCompletion(cacheMissMonitorBuildService)
+
+        gradleRootProject.extensions.extraProperties.set("cacheMiss-${rootProjectName}", cacheMissMonitorBuildService.get().cacheMiss)
+
+        if (!isInBuildLogic) { // BuildScanExtension is only available in the gradle project
```

Comment:
It seems like this should go to another plugin, given that this should only be applied to the root project, and not to build logic.

Reply:
I'd like these logic to be together in one build script so it's more readable and understandable.